modcc/solvers.cpp

```diff
@@ -171,6 +170,84 @@ static expression_ptr as_expression(symge::symbol_term_diff diff) {
     }
 }
 
+std::vector<local_assignment> SystemSolver::generate_row_updates(scope_ptr scope, std::vector<symge::symbol> row_sym) {
+    std::vector<local_assignment> S_;
+    for (const auto& s: row_sym) {
+        if (primitive(s)) continue;
+
+        auto expr = as_expression(definition(s));
+        auto local_t_term = make_unique_local_assign(scope, expr.get(), "t_");
+        auto t_ = local_t_term.id->is_identifier()->spelling();
+        symtbl_.name(s, t_);
+        S_.push_back(std::move(local_t_term));
+    }
+    return S_;
+}
+
+local_assignment SystemSolver::generate_normalizing_term(scope_ptr scope, std::vector<symge::symbol> row_sym) {
+    // Get the max element in the row
+    expression_ptr max;
+    for (auto &s: row_sym) {
+        auto elem = make_expression<IdentifierExpression>(Location(), symtbl_.name(s));
+        auto abs = make_expression<AbsUnaryExpression>(elem->location(), elem->is_identifier()->clone());
+        if (!max) {
+            max = std::move(abs);
+        } else {
+            max = make_expression<MaxBinaryExpression>(elem->location(), max->clone(), std::move(abs));
+        }
+    }
+    // Safely invert max
+    auto inv = make_expression<SafeInvUnaryExpression>(max->location(), std::move(max));
+
+    // Create a symbol for inv
+    auto local_inv_term = make_unique_local_assign(scope, inv.get(), "inv_");
+
+    return std::move(local_inv_term);
```

Comment:
gcc just warned me that the `std::move` here prevents copy-elision; we can just say `return local_inv_term;`.

Reply:
Okay, I'll do that